### PR TITLE
Optimize the update of activities

### PR DESCRIPTION
**Pull request by Walfar**
Instead of clearing all markers on map and refetching all activities from DB, we listen for changes in the DB, and only update the marker corresponding to the activity that changed.
- Added a snapshot listener on document in FirestoreActivityManager
- Added a HashMap associating an activity id to a marker, to pick the exact activity that was updated and remove/add the corresponding marker
- When fetching an activity: add a listener to the doc, then put the activity and its marker in the HashMap
- When changes is detected, remove marker and add a new one with a tag associated to the updated activity
- Added permission checks for gallery intent when editing picture

### CODE REVIEW

File: app/src/androidTest/java/com/sdp/movemeet/view/map/MainMapFragmentTest.java [hunk 1]
```diff
@@ -171,7 +171,7 @@ public void clickingOnActivityMakesCorrectIntentWhenLogged() {
             public void run() {
                 mapFragment.displayNearbyMarkers();
                 if (mapFragment.googleMap != null) {
-                    mapFragment.onMarkerClick(mapFragment.activitiesMarkers.get(0));
+                    mapFragment.onMarkerClick(mapFragment.activitiesMarkerMap.get(act));
```
Comment (Antho1426) [inline app/src/androidTest/java/com/sdp/movemeet/view/map/MainMapFragmentTest.java:174]: Is this change (.get(0) --> .get(act)) related to your optimisation for fetching *only* the updated activity-ies?

File: app/src/main/java/com/sdp/movemeet/utility/ActivitiesUpdater.java [hunk 2]
```diff
@@ -70,11 +74,13 @@ public static void updateListActivities(OnCompleteListener listener) {
      *
      * @param task task containing the requested snapshots, that would then be deserialized into activities
      */
-    private static Task addActivitiesOnSuccess(Task<QuerySnapshot> task) {
+    private static Task addActivitiesOnSuccess(Task<QuerySnapshot> task, EventListener<DocumentSnapshot> eventListener) {
```
Comment (Antho1426) [inline app/src/main/java/com/sdp/movemeet/utility/ActivitiesUpdater.java:77]: I guess this is the place where you implemented your listener to listen to updated activities. Good that you had the time to implement this!

File: app/src/main/java/com/sdp/movemeet/view/profile/EditProfileActivity.java [hunk 3]
```diff
@@ -109,8 +111,10 @@ private void assignViewsAndAdjustData() {
 
 
     public void changeProfilePicture(View view) {
-        Intent openGalleryIntent = new Intent(Intent.ACTION_PICK, MediaStore.Images.Media.EXTERNAL_CONTENT_URI);
-        startActivityForResult(openGalleryIntent, REQUEST_IMAGE);
+        if (isStorageReadPermissionGranted(this)) {
```
Comment (Antho1426) [inline app/src/main/java/com/sdp/movemeet/view/profile/EditProfileActivity.java:130]: Nice, thanks for adding this! As discussed, otherwise we could access the gallery to select a profile picture by default, without any permission...

**Review by Antho1426 — APPROVED** [review]

File: app/src/androidTest/java/com/sdp/movemeet/view/map/MainMapFragmentTest.java (hunk 1 above)
Comment (Walfar) [reply app/src/androidTest/java/com/sdp/movemeet/view/map/MainMapFragmentTest.java:174]: At first, I implemented a list of markers in the map fragment, only in the interest of making this test work (as I can't have access to a marker directly from the map). But, with the new implementation, I had to associate each activities with their respective markers, so that I could only update the marker of the activity that changed. Doing that as a HashMap, I figured that I could directly use this map instead of a list in this test, killing two birds with one stone.

Comment (Walfar) [reply app/src/androidTest/java/com/sdp/movemeet/view/map/MainMapFragmentTest.java:174]: At first, I implemented a list of markers in the map fragment, only in the interest of making this test work (as I can't have access to a marker directly from the map). But, with the new implementation, I had to associate each activities with their respective markers, so that I could only update the marker of the activity that changed. Doing that as a HashMap, I figured that I could directly use this map instead of a list in this test, killing two birds with one stone.

File: app/src/main/java/com/sdp/movemeet/utility/ActivitiesUpdater.java (hunk 2 above)
Comment (Walfar) [reply app/src/main/java/com/sdp/movemeet/utility/ActivitiesUpdater.java:77]: Yes ! Maybe this isn't the smoothest way to do it, but I didn't want to change too much of the already implemented architecture.

File: app/src/main/java/com/sdp/movemeet/view/profile/EditProfileActivity.java (hunk 3 above)
Comment (Walfar) [reply app/src/main/java/com/sdp/movemeet/view/profile/EditProfileActivity.java:130]: True, although I still find it weird that android studio didn't complain about the version without checks...

File: app/src/main/java/com/sdp/movemeet/utility/ActivityPictureCache.java [hunk 4]
```diff
@@ -28,9 +28,6 @@
 import java.io.File;
 import java.io.FileOutputStream;
 
-/**
- * This utility class is used to load/save the activities header pictures from/to cache
- */
```
Comment (bitshrine) [inline app/src/main/java/com/sdp/movemeet/utility/ActivityPictureCache.java:33]: Why did you delete this bit of documentation?

**Review by bitshrine — APPROVED** [review]
